### ISSUE TITLE
MultiDistributor: Optimize rewards data struct

### ISSUE THREAD

**Issue by facuspagnuolo** [balancer-labs/balancer-v2-monorepo#897]
The reward struct uses `uint256` for all its items. It can be optimized as follows for instance:

```solidity
struct Reward {
  uint64 rewardsDuration;
  uint64 periodFinish;
  uint64 lastUpdateTime;
  uint128 rewardRate;
  uint128 rewardPerTokenStored;
}
```

Note that the three timestamps are long enough for this use case, it would allow defining periods up to ~584 billion years which is nonsense. The same for the reward rates, additionally this last ones end up in an amount to be handled by the Vault, which cannot handle amounts bigger than `2^112 - 1`.

**Comment by nventuro** [comment]
For reference re. `rewardRate` (which is a fixed-point value), 2**128 / 10**18 yields ~3.4*10**20, so we'd be able to create distributions where 10**20 tokens are being sent per second, even if we introduced no scaling.

**Comment by nventuro** [comment]
Closing as the milestone itself is being closed.
